core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1261,6 +1270,44 @@ class KafkaController(val config: KafkaConfig,
     controlledShutdownCallback(controlledShutdownResult)
   }
 
+  private val topicConfigCache = new BoundedTimedCache[String, Properties](
+    config.controlledShutdownTopicConfigCacheSize,
+    Duration.ofMillis(config.controlledShutdownTopicConfigCacheTTLMs))
+
+  private def fetchTopicConfig(topicName: String): Properties = {
+    val cachedConfig = topicConfigCache.get(topicName)
+    if (cachedConfig != null) {
+      return cachedConfig
+    }
+
+    val config = zkClient.getEntityConfigs(ConfigType.Topic, topicName)
+    topicConfigCache.put(topicName, config)
+    return config
+  }
+
+  private def safeToShutdown(id: Int, brokerEpoch: Long): Boolean = {
+    // If a topic doesn't have min.insync.replicas configured, default to 1
+    val defaultMinISRPropertyValue = "1"
+
+    val atRiskPartitions = controllerContext.partitionsOnBroker(id).filter { partition =>
+      // Look up minISR for this topic, or use the default if not configured.
+      val minISR: Int = fetchTopicConfig(partition.topic())
```

Comment:
This still fetches the topic's configs sequentially. When the topicConfigCache is cold, and a given broker is likely hosting a partition for every topic, this could possibly take a very long time and thus choke the controller.

Reply:
Yeah I was worried about that. What do you think we should do to fix this?

One option is to scan and watch all topic configuration controller startup. That way at least it doesn't pause while the controller is handling requests.

Another I guess is to have a thread pool for gathering topic configuration to add some parallelism, which would speed this up for the current code where it is done in the ControlledShutdownRequest handler *or* during controller startup.

Another thing is that we don't necessarily have to cache all topic configuration in memory like this. We could just store the min.insync.replicas setting in a map.

Reply:
DynamicConfigManager keeps check of topic config changes. could we leverage DynamicConfigManager here so that whenever there is a topic config change,  kafka controller can be updated for the change.

Reply:
Oh that makes this really easy, thanks xiowu0.

Reply:
I updated TopicConfigHandler (owned by DynamicConfigManager) just update a map of topic -> min.insync.replicas in ControllerContext and deleted the cache.